app/exporters/patients_with_history_exporter.rb

```diff
@@ -0,0 +1,204 @@
+          "BP #{i} Medication 5",
+          "BP #{i} Dosage 5",
+          "BP #{i} Other Medications"]
+      end,
+      "Latest Blood Sugar Date",
+      "Latest Blood Sugar Value",
+      "Latest Blood Sugar Type"].flatten.compact
+  end
+
+  def self.csv_fields(patient)
+    registration_facility = patient.registration_facility
+    latest_bps = patient.latest_blood_pressures.first(DISPLAY_BLOOD_PRESSURES)
+    latest_blood_sugar = patient.latest_blood_sugar
+    latest_appointment = patient.latest_scheduled_appointment
+    latest_bp_passport = patient.latest_bp_passport
+    cache_medication_history(patient, latest_bps.map(&:recorded_at))
+    zone_column_index = csv_headers.index(zone_column)
+
+    csv_fields = [
+      patient.recorded_at.presence && I18n.l(patient.recorded_at),
+      patient.recorded_at.presence && quarter_string(patient.recorded_at),
+      ("Died" if patient.status == "dead"),
+      patient.id,
+      latest_bp_passport&.shortcode,
+      patient.full_name,
+      patient.current_age,
+      patient.gender.capitalize,
+      patient.phone_numbers.last&.number,
+      patient.address.street_address,
+      patient.address.village_or_colony,
+      patient.address.district,
+      patient.address.state,
+      registration_facility&.name,
+      registration_facility&.facility_type,
+      registration_facility&.district,
+      registration_facility&.state,
+      ("High" if patient.high_risk?),
+      latest_appointment&.days_overdue,
+      (1..DISPLAY_BLOOD_PRESSURES).map do |i|
+        bp = latest_bps[i - 1]
+        previous_bp = latest_bps[i]
+        appointment = appointment_created_on(patient, bp&.recorded_at)
+
+        [bp&.recorded_at.presence && I18n.l(bp&.recorded_at),
+          bp&.recorded_at.presence && quarter_string(bp&.recorded_at),
+          bp&.systolic,
+          bp&.diastolic,
+          bp&.facility&.name,
+          bp&.facility&.facility_type,
+          bp&.facility&.district,
+          bp&.facility&.state,
+          appointment&.facility&.name,
+          appointment&.scheduled_date,
+          appointment&.follow_up_days,
+          medication_updated?(patient, bp&.recorded_at, previous_bp&.recorded_at),
+          *medications_for(patient, bp&.recorded_at)]
+      end,
+      latest_blood_sugar&.recorded_at.presence && I18n.l(latest_blood_sugar&.recorded_at),
+      latest_blood_sugar&.to_s,
+      blood_sugar_type(latest_blood_sugar)
+    ].flatten
+
+    csv_fields.insert(zone_column_index, patient.address.zone) if zone_column_index
+    csv_fields
+  end
+
+  class << self
+    private
+
+    def zone_column
+      "Patient #{Address.human_attribute_name :zone}"
+    end
+
+    def appointment_created_on(patient, date)
+      patient.appointments
+        .where(device_created_at: date&.all_day)
+        .order(device_created_at: :asc)
+        .first
+    end
+
+    def cache_medication_history(patient, dates)
+      @medications = {patient => {}}
```

Comment:
Why store a hash with a patient key if this method only deals with one patient?

Is the intention to build up a hash of several patients? In that case, should we rewrite this as:

```
@medications ||= {}
@medications[patient] ||= {}
```

Right now, we appear to be clobbering the hash each time invoke the method.

Reply:
This fetches and keeps medications for each patient only, since we populate the CSV from latest -> oldest but the `medication updated?` calculation needs to refer to older medications while populating later medications. This is just to avoid double querying. Maybe a better name would be `fetch_medication_history`?

Reply:
That may be a better name indeed. If there's only a single key we're dealing, perhaps we don't need the `patient` key at all then.

```
{
  date_1: [stuff],
  date_2: [stuff]
}
```